database/src/ops/block.rs

```diff
@@ -1,89 +1,492 @@
+                    tables.rct_outputs_mut(),
+                )?;
+            }
+        }
+    }
+
+    Ok(())
 }
 
-/// TODO
-pub fn get_block_cumulative_rct_outputs() {
-    todo!()
+//---------------------------------------------------------------------------------------------------- `pop_block_*`
+/// Remove the top/latest block from the database.
+///
+/// The removed block's height and hash are returned.
+#[doc = doc_error!()]
+// no inline, too big
+pub fn pop_block(tables: &mut impl TablesMut) -> Result<(BlockHeight, BlockHash), RuntimeError> {
+    // Remove block data from tables.
+    let (block_height, block_hash) = {
+        let (block_height, block_info) = tables.block_infos_mut().pop_last()?;
+        (block_height, block_info.block_hash)
+    };
+
+    // Block heights.
+    tables.block_heights_mut().delete(&block_hash)?;
+
+    // Block blobs.
+    // We deserialize the block blob into a `Block`, such
+    // that we can remove the associated transactions later.
+    let block_blob = tables.block_blobs_mut().take(&block_height)?.0;
+    let block = Block::read(&mut block_blob.as_slice())?;
+
+    // Transaction & Outputs.
+    for tx_hash in block.txs {
+        let (tx_id, tx_blob) = remove_tx(&tx_hash, tables)?;
+        let tx = Transaction::read(&mut tx_blob.0.as_slice())?;
+
+        // Is this a miner transaction?
+        let mut miner_tx = false;
+        for inputs in &tx.prefix.inputs {
+            match inputs {
+                // Key images.
+                Input::ToKey { key_image, .. } => {
+                    let result =
+                        remove_key_image(key_image.compress().as_bytes(), tables.key_images_mut());
+
+                    // TODO: all test tx's have the same key image so the 1st
+                    // removal removes everything - fix me after we have real data.
+                    if cfg!(test) {
+                        match result {
+                            Ok(()) | Err(RuntimeError::KeyNotFound) => (),
+                            Err(e) => return Err(e),
+                        }
+                    } else {
+                        result?;
+                    }
+                }
+                // This is a miner transaction, set it for later use.
+                Input::Gen(_) => miner_tx = true,
+            }
+        }
+
+        // Remove each output in the transaction.
+        for (i, output) in tx.prefix.outputs.into_iter().enumerate() {
+            // Outputs with clear amounts.
+            if let Some(amount) = output.amount {
+                // RingCT miner outputs.
+                if miner_tx && tx.prefix.version == 2 {
+                    let amount_index = get_rct_num_outputs(tables.rct_outputs())?.saturating_sub(1);
+                    remove_rct_output(&amount_index, tables.rct_outputs_mut())?;
+                // Pre-RingCT outputs.
+                } else {
+                    let amount_index = tables.num_outputs_mut().take(&amount)?;
```

Comment:
We need a minus 1 here right?

Reply:
Same question as https://github.com/Cuprate/cuprate/pull/102#discussion_r1568076162

Reply:
This is removing outputs so the index of the top output is one less than the amount of outputs

Reply:
@hinto-janai This should be get not take right? and we still need the minus 1